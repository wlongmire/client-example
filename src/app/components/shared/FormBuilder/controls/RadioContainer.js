import React, { PropTypes } from 'react'

import {
  FormGroup,
  Radio,
  ControlLabel,
  Tooltip,
  OverlayTrigger
} from 'react-bootstrap'
import AdditionalInfoComponent from './AdditionalInfoComponent'

class RadioContainer extends React.Component {
  constructor(props) {
    super(props)

    this.onChangeHandler = this.onChangeHandler.bind(this)
    const name = this.props.data.name

    this.state = {
      // value: this.props.initialValues[name] ? this.props.initialValues[name] : null,
      disabled: (this.props.initialParams[name] && this.props.initialParams[name].disabled) ?
        this.props.initialParams[name].disabled : false,
      options: [].concat(this.props.data.attributes.options),
      isValid: null,
      validationMessage: ''
    }
  }

  componentWillMount() {
    if (this.props.initialValues[this.props.data.name]) {
      const event = {
        target: {
          value: this.props.initialValues[this.props.data.name]
        }
      }
      this.onChangeHandler(event)
    } else {
      this.setState({
        value: null
      })
    }
  }

  getValidationState() {
    // this bit of code acts as a circuit breaker
    // We want get validation to trigger a validation function
    // and when results are back we need to call it again
    // But this breaks the loops
    if (this.state.isValid) {
      const isValid = this.state.isValid
      this.state.isValid = null
      return isValid
    }

    // only trigger validation if the value changes
    if (this.state.value !== '' &&
        this.props.data.attributes &&
        this.props.data.attributes.validationFunc &&
        this.props.validation[this.props.data.attributes.validationFunc]) {
      this.props.validation[this.props.data.attributes.validationFunc](this.props.data.name, this.state.value).then((result) => {
        this.setState({
          isValid: (result) ? 'success' : 'error',
          validationMessage: (result.status) ? '' : result.message
        })
      })
    }

    if (this.state.isValid !== null) {
      return this.state.isValid
    }

    if (this.props.data.attributes && this.props.data.attributes.validationRegEx) {
      const regex = new RegExp(unescape(this.props.data.attributes.validationRegEx))
      return (regex.test(this.state.value)) ? 'success' : 'error'
    }

    return
  }

  onChangeHandler(event) {
    // get the data for selected option
    const option = this.props.data.attributes.options.filter((optionResult) => {
      return optionResult.value.toString() === event.target.value
    })[0]

    // trigger supplemental questions
    if (option.supplementalquestionIds && option.supplementalquestionIds.length > 0) {
      this.props.handleSupplementTrigger(option.supplementalquestionIds)
    } else {
      this.props.handleSupplementTrigger([])
    }

    // set value to selected option
    this.setState({
      value: event.target.value
    })

    this.props.handleFormChange()
  }

  render() {
    const tooltip = (<Tooltip id={`tooltip_${this.props.data.questionId}`}> {this.props.data.tooltiptext}</Tooltip>)
    const optionsItems = this.props.data.attributes.options || []

    const options = optionsItems.map((data, index) => {
      const checked = ((this.state.value) && (this.state.value.toString() === data.value.toString()))

      return (
        <Radio
          className={this.props.data.name}
          key={index}
          onChange={this.onChangeHandler}
          value={data.value}
          checked={checked}
          inline={this.props.data.verticalRadioAlign}
          disabled={this.state.disabled}
          name={`optionGroup_${this.props.data.questionId}`}
        >
          <span className="radioText">{data.text}</span>
        </Radio>
      )
    })

    return (
      <FormGroup
        className="radioGroup"
        validationState={this.getValidationState()}
        controlId={this.props.data.name}
      >
        {/* I'm adding the overlay to the ControlLabel in the case of Radio buttons.
        having a tooltip on the Radio buttons felt a little counter-intuitive in terms of UX*/}
        { this.props.data.text && <OverlayTrigger placement="top" overlay={tooltip} trigger={(this.props.data.tooltiptext) ? ['hover', 'focus'] : null}>
          <ControlLabel>{this.props.data.text}</ControlLabel>
        </OverlayTrigger> }
        <div className="options">
          {options}
        </div>
        <AdditionalInfoComponent
          additionalInfo1Color={this.props.data.additionalInfo1Color}
          additionalInfo1={this.props.data.additionalInfo1}
          additionalInfo2={this.props.data.additionalInfo2}
        />
      </FormGroup>
    )
  }
}


RadioContainer.propTypes = {
  data: PropTypes.object,
  initialParams: PropTypes.object,
  initialValues: PropTypes.object,
  validation: PropTypes.object,
  handleSupplementTrigger: PropTypes.func,
  handleFormChange: PropTypes.func
}


export default RadioContainer