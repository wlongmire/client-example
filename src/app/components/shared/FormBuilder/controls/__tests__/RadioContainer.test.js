/* eslint-env node, jest */

import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { RadioContainer } from '../RadioContainer'

const dataProps = {
  attributes: {
    options: [
      { optionId: '1', text: 'Yes', value: true },
      { optionId: '2', text: 'No', value: false },
    ]
  },
  inputType: 'radio',
  name: 'occupancy',
  questionId: '11',
  required: true,
  text: 'Will there be occupancy during the project?'
}

describe('>>> RadioContainer Container Component --- Snapshot', () => {
  it('+++ provide a radio input field', () => {
    const renderedValue = shallow(
      <RadioContainer
        data={dataProps}
        handleSupplementTrigger={jest.fn}
        handleFormChange={jest.fn}
        initialValues={{}}
        initialParams={{}}
      />)
    expect(toJson(renderedValue)).toMatchSnapshot()
  })
})