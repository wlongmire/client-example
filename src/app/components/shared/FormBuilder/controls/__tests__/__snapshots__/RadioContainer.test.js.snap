// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> RadioContainer Container Component --- Snapshot +++ provide a radio input field 1`] = `
<FormGroup
  bsClass="form-group"
  className="radioGroup"
  controlId="occupancy"
>
  <OverlayTrigger
    defaultOverlayShown={false}
    overlay={
      <Tooltip
        bsClass="tooltip"
        id="tooltip_11"
        placement="right"
      >
         
      </Tooltip>
    }
    placement="top"
    trigger={null}
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Will there be occupancy during the project?
    </ControlLabel>
  </OverlayTrigger>
  <div
    className="options"
  >
    <Radio
      bsClass="radio"
      checked={null}
      className="occupancy"
      disabled={false}
      inline={false}
      name="optionGroup_11"
      onChange={[Function]}
      value={true}
    >
      <span
        className="radioText"
      >
        Yes
      </span>
    </Radio>
    <Radio
      bsClass="radio"
      checked={null}
      className="occupancy"
      disabled={false}
      inline={false}
      name="optionGroup_11"
      onChange={[Function]}
      value={false}
    >
      <span
        className="radioText"
      >
        No
      </span>
    </Radio>
  </div>
  <AdditionalInfoComponent />
</FormGroup>
`;
